File: 2023/day19/part2.py
import sys
import copy

input_filepath = sys.argv[1]

lines = open(input_filepath).read().strip().split("\n")
L = []
i = 0
while len(lines[i]) > 0:
    L.append(lines[i])
    i += 1


class Ruleset:
    def __init__(self, nextStep):
        self.rules = []
        self.nextStep = nextStep


# build hashmap
workflows = {}
for line in L:
    name = line[: line.index("{")]
    rules = line[line.index("{") + 1 : line.index("}")]

    workflows[name] = rules.split(",")

acceptable_rulesets = []
move_stack = [Ruleset("in")]

while len(move_stack) > 0:
    ruleset = copy.deepcopy(move_stack.pop())

    # if the next step is "A", add to acceptable, if it's R, do nothing
    if ruleset.nextStep == "A":
        acceptable_rulesets.append(copy.deepcopy(ruleset))
        continue
    if ruleset.nextStep == "R":
        continue

    workflow = workflows[ruleset.nextStep]

    # if we are at the end, then add just one thing
    if len(workflow) == 1:
        # remove from hashmap since we don't need it anymore
        workflows.pop(ruleset.nextStep)

        # add next step to ruleset
        ruleset.nextStep = workflow[0]
        move_stack.append(ruleset)

        continue

    line = workflow[0]

    # pare down the workflow
    workflows[ruleset.nextStep] = workflows[ruleset.nextStep][1:]

    # now add both ways to the ruleset
    next_rule, next_workflow = line.split(":")

    # have to add it both ways
    satisfied_ruleset = copy.deepcopy(ruleset)
    satisfied_ruleset.rules.append(next_rule)
    satisfied_ruleset.nextStep = next_workflow
    move_stack.append(satisfied_ruleset)

    # now add the other way (not satisfied )
    if ">" in next_rule:
        next_rule = next_rule.replace(">", "<=")
    else:
        next_rule = next_rule.replace("<", ">=")

    ruleset.rules.append(next_rule)
    move_stack.append(ruleset)

# define a function to deal with an acceptable ruleset
working_parts = set()


def acceptable_part_combos(ruleset):
    vals = {
        "x": list(range(1, 4001)),
        "m": list(range(1, 4001)),
        "a": list(range(1, 4001)),
        "s": list(range(1, 4001)),
    }

    for rule in ruleset.rules:
        dict_val = rule[0]
        if "<=" in rule:
            operator = "<="
            val = int(rule[3:])
        elif ">=" in rule:
            operator = ">="
            val = int(rule[3:])
        else:
            operator = rule[1]
            val = int(rule[2:])

        if operator == ">":
            vals[dict_val] = [x for x in vals[dict_val] if x > val]
        elif operator == ">=":
            vals[dict_val] = [x for x in vals[dict_val] if x >= val]
        elif operator == "<=":
            vals[dict_val] = [x for x in vals[dict_val] if x <= val]
        else:
            vals[dict_val] = [x for x in vals[dict_val] if x < val]

    result = 1
    for val in vals.values():
        result *= len(val)
        # print(min(val), max(val))
    # print(ruleset.rules, result)
    return result


# first need to collapse down the ruleset
res = 0
for ruleset in acceptable_rulesets:
    res += acceptable_part_combos(ruleset)
print(res)

File: 2023/day19/Part2.java

import java.util.*;
import java.io.*;

public class Part2 {
  public static void main(String[] args) {
    if (args.length != 1) {
      System.out.println("Usage: <input_filepath");
    }

    String input_filepath = args[0];
    BufferedReader br;
    ArrayList<String> lines = new ArrayList<>();

    try {
      br = new BufferedReader(new FileReader(input_filepath));
      String line;
      while ((line = br.readLine()) != null) {
        lines.add(line);
      }
    } catch (IOException e) {
      System.out.println(e);
      return;
    }

    // construct the workflows
    int i = 0;
    HashMap<String, Workflow> workflows = new HashMap<>();
    while (true) {
      if (lines.get(i).length() == 0) {
        i++;
        break;
      }

      String line = lines.get(i);
      Workflow workflow = new Workflow(line);
      workflows.put(workflow.name, workflow);
      i++;
    }

    long res = 0;

    // for each workflow, we are going to figure out the range of values that can
    // lead us there

    System.out.println(res);
  }
}

class Rule {
  public String partName;
  public String operator;
  public Integer value;
  public String destination;

  public Rule(String line) {
    String[] splitLine = line.split(":");
    destination = splitLine[1];
    line = splitLine[0];

    if (line.contains("<")) {
      operator = "<";
      partName = line.split("<")[0];
      value = Integer.parseInt(line.split("<")[1]);
    } else {
      operator = ">";
      partName = line.split(">")[0];
      value = Integer.parseInt(line.split(">")[1]);
    }
  }
}

class Workflow {
  public ArrayList<Rule> rules;
  public String name;
  public String lastStep;

  public Workflow(String line) {
    rules = new ArrayList<>();
    name = line.substring(0, line.indexOf("{"));

    // substring to get rid of stuff we don't need
    line = line.substring(line.indexOf("{") + 1, line.indexOf("}"));
    String[] stringRules = line.split(",");
    for (int i = 0; i < stringRules.length; i++) {
      if (i == stringRules.length - 1) {
        lastStep = stringRules[i];
        continue;
      }

      rules.add(new Rule(stringRules[i]));
    }
  }

  public String processWorkflow(PartCollection pc) {
    for (Rule rule : rules) {
      if (rule.operator.equals(">")) {
        if (pc.partValues.get(rule.partName) > rule.value) {
          return rule.destination;
        }
      } else if (rule.operator.equals("<")) {
        if (pc.partValues.get(rule.partName) < rule.value) {
          return rule.destination;
        }
      }
    }

    return lastStep;
  }
}

class PartCollection {
  public HashMap<String, Integer> partValues;
  public Integer partSum;

  public PartCollection(String line) {
    line = line.replace("{", "");
    line = line.replace("}", "");

    partValues = new HashMap<>();

    String[] splitLine = line.split(",");
    partValues.put("x", Integer.parseInt(splitLine[0].split("=")[1]));
    partValues.put("m", Integer.parseInt(splitLine[1].split("=")[1]));
    partValues.put("a", Integer.parseInt(splitLine[2].split("=")[1]));
    partValues.put("s", Integer.parseInt(splitLine[3].split("=")[1]));

    partSum = 0;
    for (Integer val : partValues.values()) {
      partSum += val;
    }
  }
}
